File: 3764-maximum-sum-with-at-most-k-elements/3764-maximum-sum-with-at-most-k-elements.java
class Solution {
    public long maxSum(int[][] grid, int[] limits, int k) {
        long ans = 0;
        int n = grid.length, m = grid[0].length;
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);

        for (int i = 0; i < n; ++i) {
            Arrays.sort(grid[i]);
            for (int j = m - 1; j >= 0 && limits[i] > 0; j--) {
                pq.offer(grid[i][j]);
                limits[i]--;
            }
        }
        while (k-- > 0 && !pq.isEmpty())
            ans += pq.poll();
        
        return ans;
    }
}

File: 3764-maximum-sum-with-at-most-k-elements/3764-maximum-sum-with-at-most-k-elements.cpp
class Solution {
public:
    long long maxSum(vector<vector<int>>& grid, vector<int>& limits, int k) {
        long long ans = 0;
        int n = grid.size(), m = grid[0].size();

        priority_queue<int, vector<int>> pq;

        for (int i = 0; i < n; ++i) {
            sort(grid[i].rbegin(), grid[i].rend());
            for (int j = 0; j < m and limits[i]; j++)
                pq.push(grid[i][j]), limits[i]--;
        }
        while (k-- and !pq.empty()) {
            ans += pq.top();
            pq.pop();
        }
        return ans;
    }
};